File: sample/todo-app-ios/todo-app-ios/LifecycleRegistryExt.swift
//
//  LifecycleRegistryExt.swift
//  todo-app-ios
//
//  Created by Arkadii Ivanov on 5/12/20.
//  Copyright © 2020 arkivanov. All rights reserved.
//

import TodoLib

extension LifecycleRegistry {
    
    func resume() {
        if (state == .initialized) {
            onCreate()
        }

        if (state == .created) {
            onStart()
        }

        if (state == .started) {
            onResume()
        }
    }
    
    func stop() {
        if (state == .resumed) {
            onPause()
        }

        if (state == .started) {
            onStop()
        }
    }
    
    func destroy() {
        stop()
        
        if (state == .created) {
            onDestroy()
        }
    }
}

File: mvikotlin/src/commonMain/kotlin/com/arkivanov/mvikotlin/core/lifecycle/LifecycleRegistryExt.kt
package com.arkivanov.mvikotlin.core.lifecycle

import com.arkivanov.mvikotlin.core.lifecycle.Lifecycle.State

fun LifecycleRegistry.resume() {
    if (state == State.INITIALIZED) {
        onCreate()
    }

    if (state == State.CREATED) {
        onStart()
    }

    if (state == State.STARTED) {
        onResume()
    }
}

fun LifecycleRegistry.stop() {
    if (state == State.INITIALIZED) {
        onCreate()
    } else {
        if (state == State.RESUMED) {
            onPause()
        }

        if (state == State.STARTED) {
            onStop()
        }
    }
}

fun LifecycleRegistry.destroy() {
    stop()

    if (state == State.CREATED) {
        onDestroy()
    }
}
